cpp/src/arrow/buffer.h

```diff
@@ -254,20 +252,51 @@ class ARROW_EXPORT Buffer {
   /// `data()` pointer.  Otherwise, you'll have to `View()` or `Copy()` it.
   bool is_cpu() const { return is_cpu_; }
 
+  /// \brief Whether the buffer is mutable
+  ///
+  /// If this function returns true, you are allowed to modify buffer contents
+  /// using the pointer returned by `mutable_data()` or `mutable_address()`.
+  bool is_mutable() const { return is_mutable_; }
+
   const std::shared_ptr<Device>& device() const { return memory_manager_->device(); }
 
   const std::shared_ptr<MemoryManager>& memory_manager() const { return memory_manager_; }
 
   std::shared_ptr<Buffer> parent() const { return parent_; }
 
-  // Convenience functions
+  /// \brief Get a RandomAccessFile for reading a buffer
+  ///
+  /// The returned file object reads from this buffer's underlying memory.
   static Result<std::shared_ptr<io::RandomAccessFile>> GetReader(std::shared_ptr<Buffer>);
+
+  /// \brief Get a OutputStream for writing to a buffer
+  ///
+  /// The buffer must be mutable.  The returned stream object writes into the buffer's
+  /// underlying memory (but it won't resize it).
   static Result<std::shared_ptr<io::OutputStream>> GetWriter(std::shared_ptr<Buffer>);
 
+  /// \brief Copy buffer
+  ///
+  /// The buffer contents will be copied into a new buffer allocated by the
+  /// given MemoryManager.  This function supports cross-device copies.
   static Result<std::shared_ptr<Buffer>> Copy(std::shared_ptr<Buffer> source,
                                               const std::shared_ptr<MemoryManager>& to);
+
+  /// \brief View buffer
+  ///
+  /// Return a Buffer that reflects this buffer, seen potentially from another device,
+  /// without explicitly copying its contents.  How this is implemented depends
```

Comment:
```suggestion
  /// without necessarily copying its contents.  How this is implemented depends
```
?

Reply:
Hmm, I should explain what I mean in that docstring. If you "view" a GPU buffer on the CPU, you call a NVidia API which gives you the CPU-readable address of the GPU buffer. But, of course, when you do the actual reading, _something_ has to happen: either the data is transferred byte per byte on the PCIe bus (probably slow?), or the data is cached in some host memory area for faster retrieval. But that caching is implicit and not seen by userspace (hence "without explicitly copying its contents").

With that in mind, do you still think the wording should be changed?

Reply:
Hmmm, maybe:
"Concrete devices and buffer kinds may support transfer mechanisms more sophisticated than an an explicit eager copy. For example `CudaBuffer`s may be viewed by a CPU device directly: data will be transferred from the GPU transparently when the CPU buffer is accessed.

\return An equivalent Buffer viewed from the provided memory manager's device, or nullptr if only eager copy is supported."

Reply:
Hmm, not terribly fond of that spelling :-) I'm trying out another one.